### PR TITLE
Implement migration sequencing (phase 1)

### PR DESCRIPTION
**Pull request by ericvergnaud**
## Changes
Users need the ability to sequence migration tasks
This PR implements sequencing for workflow tasks, jobs and clusters.
It will be followed by PRs for other workspace items 

### Linked issues
Progresses #1415 

### Functionality
None

### Tests
- [x] added unit tests

### CODE REVIEW

File: src/databricks/labs/ucx/sequencing/sequencing.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+            node_id=0, object_type="ROOT", object_id="ROOT", object_name="ROOT", object_owner="NONE"
+        )
+
+    def register_workflow_task(self, task: jobs.Task, job: jobs.Job, _graph: DependencyGraph) -> MigrationNode:
+        task_id = f"{job.job_id}/{task.task_key}"
+        task_node = self._find_node(object_type="TASK", object_id=task_id)
+        if task_node:
+            return task_node
+        job_node = self.register_workflow_job(job)
+        MigrationNode.last_node_id += 1
```
Comment (nfx) [inline src/databricks/labs/ucx/sequencing/sequencing.py:82]:
```suggestion
        self._last_node_id += 1
```
make it a private property of this class

File: src/databricks/labs/ucx/sequencing/sequencing.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+class MigrationSequencer:
+
+    def __init__(self, ws: WorkspaceClient):
+        self._ws = ws
+        self._root = MigrationNode(
+            node_id=0, object_type="ROOT", object_id="ROOT", object_name="ROOT", object_owner="NONE"
+        )
+
+    def register_workflow_task(self, task: jobs.Task, job: jobs.Job, _graph: DependencyGraph) -> MigrationNode:
```
Comment (nfx) [inline src/databricks/labs/ucx/sequencing/sequencing.py:76]:
```suggestion
    def _register_workflow_task(self, task: jobs.Task, job: jobs.Job, graph: DependencyGraph) -> MigrationNode:
```
make this one private. we should start from the job, not the task

File: src/databricks/labs/ucx/sequencing/sequencing.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+                return found
+        return None
+
+
+class MigrationSequencer:
+
+    def __init__(self, ws: WorkspaceClient):
+        self._ws = ws
+        self._root = MigrationNode(
+            node_id=0, object_type="ROOT", object_id="ROOT", object_name="ROOT", object_owner="NONE"
```
Comment (nfx) [inline src/databricks/labs/ucx/sequencing/sequencing.py:73]:
```suggestion
            node_id=0, object_type="ROOT", object_id="ROOT", object_name="ROOT", object_owner="NONE",
```
`make fmt`

File: src/databricks/labs/ucx/sequencing/sequencing.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        job_node = self.register_workflow_job(job)
+        MigrationNode.last_node_id += 1
+        task_node = MigrationNode(
+            node_id=MigrationNode.last_node_id,
+            object_type="TASK",
+            object_id=task_id,
+            object_name=task.task_key,
+            object_owner=job_node.object_owner,  # no task owner so use job one
+        )
+        job_node.required_steps.append(task_node)
```
Comment (nfx) [inline src/databricks/labs/ucx/sequencing/sequencing.py:90]: this algorithm is a bit convoluted and hard to follow. please rewrite with the adjacency map and do a topological sorting with kahn's algorithm in this class. e.g.

```python
self._adjacency = collections.defaultdict(set)

...

# have a central _nodes field to have all the nodes addressable by (TYPE, ID) tuple
self._nodes[('TASK', task_id)] = MigrationNode(
            node_id=self._last_node_id,
            object_type="TASK",
            object_id=task_id,
            object_name=task.task_key,
            object_owner=job_node.object_owner,
        )
self._adjacency[('JOB', job_id)].append(('TASK', task_id))
```

... and actual toposort is pretty straightforward with Kahn's algo:

```python
indegrees = collections.defaultdict(int)
for src, dep_set in self._adjacency.items():
  indegrees[src] = len(dep_set) # count incoming dependencies for a node

queue, toposorted, sequence_num = [], [], 0
for src, incoming in indegrees.items():
  if incoming > 0: continue
  queue.append(src) # start with zero-dependencies nodes

while queue:
  curr = queue.popleft()
  toposorted.append(replace(self._nodes[curr], sequence_num=sequence_num))
  sequence_num += 1
  for dep in self._adjacency[curr]:
    indegrees[dep] -= 1
    if indegrees[dep] == 0:
      queue.append(dep)

return toposorted
```

i find it confusing with all those `_deduplicate_steps` and `_find_node` methods.

File: src/databricks/labs/ucx/sequencing/sequencing.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        if job_node:
+            return job_node
+        MigrationNode.last_node_id += 1
+        job_name = job.settings.name if job.settings and job.settings.name else str(job.job_id)
+        job_node = MigrationNode(
+            node_id=MigrationNode.last_node_id,
+            object_type="JOB",
+            object_id=str(job.job_id),
+            object_name=job_name,
+            object_owner=job.creator_user_name or "<UNKNOWN>",
```
Comment (nfx) [inline src/databricks/labs/ucx/sequencing/sequencing.py:110]: inject use `databricks.labs.ucx.framework.owners.Ownership` subclasses to properly determine object owners.

**Review by nfx — CHANGES_REQUESTED** [review]
please simplify the toposort algorithm with the provided outline and use `Ownership` subclases to get the right owner name.

File: src/databricks/labs/ucx/sequencing/sequencing.py (hunk 2 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/sequencing/sequencing.py:76]: the current thinking is to leverage the existing dependency graph in `WorkflowLinter.refresh_report` in order to avoid rebuilding it and re-fetch all assets from the workspace. Starting from the job would make that impossible.

File: src/databricks/labs/ucx/sequencing/sequencing.py (hunk 1 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/sequencing/sequencing.py:82]: done

File: src/databricks/labs/ucx/sequencing/sequencing.py (hunk 3 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/sequencing/sequencing.py:73]: `make fmt` doesn't change this code...

Comment (nfx) [reply src/databricks/labs/ucx/sequencing/sequencing.py:73]: with trailing comma it should

File: src/databricks/labs/ucx/sequencing/sequencing.py (hunk 4 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/sequencing/sequencing.py:94]: The problem with Kahn is it only works for DAGs, which we can't guarantee. We have duplicates and recursive loops. I'm trying it out but I suspect it will break on solacc.

File: src/databricks/labs/ucx/sequencing/sequencing.py (hunk 5 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/sequencing/sequencing.py]: done. Requires #2999

File: src/databricks/labs/ucx/sequencing/sequencing.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+        if job_node:
+            return job_node
+        self._last_node_id += 1
+        job_name = job.settings.name if job.settings and job.settings.name else str(job.job_id)
+        job_node = MigrationNode(
+            node_id=self._last_node_id,
+            object_type="JOB",
+            object_id=str(job.job_id),
+            object_name=job_name,
+            object_owner=JobOwnership(self._admin_locator).owner_of(job),
```
Comment (nfx) [inline src/databricks/labs/ucx/sequencing/sequencing.py:114]: inject instance of `JobOwnership` via constructor, it has to be on `GlobalContext`

File: src/databricks/labs/ucx/assessment/clusters.py [hunk 7]
```diff
@@ -192,6 +192,16 @@ def _maybe_direct_owner(self, record: ClusterInfo) -> str | None:
         return record.creator
 
 
+class ClusterDetailsOwnership(Ownership[ClusterDetails]):
+    """Determine ownership of clusters in the workspace.
+
+    This is the cluster creator (if known).
+    """
+
+    def _maybe_direct_owner(self, record: ClusterDetails) -> str | None:
```
Comment (nfx) [inline src/databricks/labs/ucx/assessment/clusters.py:201]: convert from `ClusterDetails` to `ClusterInfo` and don't do any renames/new classes. we have "cluster" as a concept within our system, `ClusterDetails` and `ClusterInfo` are just the names of API generated classes which are slightly different views on the cluster entity.

TLDR: don't rename existing ownership classes.

**Review by nfx — CHANGES_REQUESTED** [review]

File: src/databricks/labs/ucx/sequencing/sequencing.py (hunk 4 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/sequencing/sequencing.py]: Consider the following code:

file a.py:
```
class A:
 
    def return_a_bee(self):
        from b import B
        return B()

```
file b.py:

```
from a import A

class B(A):
     pass
```

the above will return the following dependency graph fragment:

    T
    / \
  A   B
 /       \
B       A

When building the dependency graph we detect the recursive cycle (and we break it).
In the above there is no 0-dependency node, so not sure Kahn will work...

File: src/databricks/labs/ucx/sequencing/sequencing.py (hunk 5 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/sequencing/sequencing.py]: #2999 no longer needed

File: src/databricks/labs/ucx/assessment/clusters.py (hunk 7 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/assessment/clusters.py]: done

File: src/databricks/labs/ucx/sequencing/sequencing.py (hunk 6 above)
Comment (ericvergnaud) [reply src/databricks/labs/ucx/sequencing/sequencing.py]: It's not on GlobalContext, and I guess making that happen deserves a dedicated PR ?
Can we make this change later ?

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,25 +1,5 @@
 # Version changelog
 
-## 0.45.0
```
Comment (nfx) [inline CHANGELOG.md:3]: bad rebase, most likely

**Review by nfx — CHANGES_REQUESTED** [review]
re-rebase

**Comment by ericvergnaud** [comment]
Superseded by #3008
